File: Behavioural/Iterator/IteratorPatternRunner.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Pattern.Behavioural.Iterator
{
	internal static class IteratorPatternRunner
	{
		public static void Run()
		{
			Console.WriteLine("Iterator");

			Console.WriteLine("Fibunacci Iterator");
			var iterator = new FibunacciIterator();

			int fib = 0;
			int max = 2048;
			while(true)
			{
				int f = iterator.Next();
				if(f > max) break;

				fib = f;
				Console.WriteLine(fib);
				Thread.Sleep(100);
			}

			Console.WriteLine("Collection iterator");
			var monthiterator = new MonthIterator();
			while(monthiterator.HasNext())
			{
				Console.WriteLine(monthiterator.Next());
			}


		}

		public interface INumberIterator
		{
			int Next();
		}

		public interface ICollectionIterator
		{
			bool HasNext();
			string Next();
		}

		public class MonthIterator : ICollectionIterator
		{
			public int Index = 0;
			public List<string> Values = new List<string>()
			{ "Januar", "Februar", "März", "April", "Mai", "Juni", 
				"Juli", "August", "September", "Oktober", "November", "Dezember"};
			public bool HasNext()
			{
				return Values.Count > Index;
			}

			public string Next()
			{
				var x = Values[Index];
				Index++;

				return x;
			}
		}
		public class FibunacciIterator : INumberIterator
		{
			public int a = 0;
			public int b = 1;

			public int Next()
			{
				int temp = a;
				a = b;
				b = temp + b;

				return a;
			}

		}
	}
}

File: VB/Behavioural/Iterator/IteratorPatternRunner.vb
Imports System.Threading

Public Class IteratorPatternRunner

	Friend Sub Run()

		Console.WriteLine("Iterator Pattern")

		Console.WriteLine("Number Iterator - Fibunacci:")
		Dim fibunacciIterator As FibunacciIterator = New FibunacciIterator(1000)
		While fibunacciIterator.HasNext()
			Dim f As UInteger = fibunacciIterator.GetNext()
			Console.WriteLine(f)

			Thread.Sleep(250)
		End While

		Console.WriteLine("Collection Iterator - Months:")
		Dim monthIterator As MonthIterator = New MonthIterator()
		While monthIterator.HasNext()
			Dim m = monthIterator.GetNext()
			Console.WriteLine(m)

			Thread.Sleep(100)
		End While

	End Sub
End Class

Interface INumberIterator
	Function HasNext() As Boolean
	Function GetNext() As Integer
End Interface

Interface ICollectionIterator
	Function HasNext() As Boolean
	Function GetNext() As String
End Interface


Friend Class FibunacciIterator
	Implements INumberIterator

	Private _a As Integer
	Private _b As Integer
	Private _max As Integer

	Sub New(ByVal max As Integer)
		_a = 0
		_b = 1
		_max = max
	End Sub

	Public Function HasNext() As Boolean Implements INumberIterator.HasNext
		If _a + _b < _max Then
			Return True
		Else
			Return False
		End If
	End Function

	Public Function GetNext() As Integer Implements INumberIterator.GetNext
		Dim c = _a + _b
		_a = _b
		_b = c

		Return c
	End Function

End Class

Friend Class MonthIterator
	Implements ICollectionIterator

	Private _index As Integer
	Private _months As List(Of String)

	Public Sub New()
		_index = 0
		_months = New List(Of String)
		_months.AddRange({"Januar", "Februar", "März", "April", "Mai", "Juni",
						 "Juli", "August", "September", "Oktober", "November", "Dezember"})

	End Sub

	Public Function HasNext() As Boolean Implements ICollectionIterator.HasNext
		Return _months.Count > _index
	End Function

	Public Function GetNext() As String Implements ICollectionIterator.GetNext
		Dim x As String = _months.ElementAt(_index)
		_index += 1

		Return x
	End Function
End Class
